### ISSUE TITLE
Release 4.0.0

### ISSUE THREAD

**Issue by manics** [jupyterhub/jupyterhub#4350]
## TODO

- [x] changelog https://github.com/jupyterhub/jupyterhub/pull/4375
- [x] singleuser extension #3888 
- [x] singleuser extension default #4354
- [x] publish a beta
- [x] announce beta
- [x] complete docs reorganization
- [ ] final release

### Proposed change

A new release of JupyterHub from `main`, including [Use XSRF tokens for cross-site checks #4032](https://github.com/jupyterhub/jupyterhub/pull/4032)

### Alternative options

Add a new section to https://jupyterhub.readthedocs.io/en/stable/admin/log-messages.html with the `Blocking Cross Origin API request. Referer: ` problem

### Who would use this feature?

People frequently hit this problem on Discourse or in the GitHub issues. https://github.com/jupyterhub/jupyterhub/pull/4032 is meant to fix it (or at least make it less likely to occur).

A secondary benefit is that we may be able to have a security review of JupyterHub done via the jupyter-security project, so would be a good chance to have the change to cross site requests verified.

### (Optional): Suggest a solution
This is the current changelog relative to 3.1.0 (3.1.1 was a backport)
```markdown
# 3.1.0...master@{2023-02-10}

([full changelog](https://github.com/jupyterhub/jupyterhub/compare/3.1.0...c93832cb33fd00cb7978a0c4428d5af9a69ba968))

## New features added

- Dynamic table for changing customizable properties of groups [#3651](https://github.com/jupyterhub/jupyterhub/pull/3651) ([@vladfreeze](https://github.com/vladfreeze))

## Enhancements made

- Standardize styling on input fields by moving common form CSS to page.less [#4294](https://github.com/jupyterhub/jupyterhub/pull/4294) ([@minrk](https://github.com/minrk))

## Bugs fixed

- Backport PR #4303 on branch 3.x (make sure event-schemas are installed) [#4317](https://github.com/jupyterhub/jupyterhub/pull/4317) ([@meeseeksmachine](https://github.com/meeseeksmachine))
- Backport PR #4302: sqlalchemy 2 compatibility [#4315](https://github.com/jupyterhub/jupyterhub/pull/4315) ([@minrk](https://github.com/minrk))
- make sure event-schemas are installed [#4303](https://github.com/jupyterhub/jupyterhub/pull/4303) ([@minrk](https://github.com/minrk))
- sqlalchemy 2 compatibility [#4302](https://github.com/jupyterhub/jupyterhub/pull/4302) ([@minrk](https://github.com/minrk))
- test and fix deprecated load_groups list [#4299](https://github.com/jupyterhub/jupyterhub/pull/4299) ([@minrk](https://github.com/minrk))
- Fix formatting of load_groups help string [#4295](https://github.com/jupyterhub/jupyterhub/pull/4295) ([@minrk](https://github.com/minrk))
- Remove remnants of unused jupyterhub-services cookie [#4258](https://github.com/jupyterhub/jupyterhub/pull/4258) ([@minrk](https://github.com/minrk))

## Maintenance and upkeep improvements

- temporary fix: pin base-notebook tag [#4348](https://github.com/jupyterhub/jupyterhub/pull/4348) ([@minrk](https://github.com/minrk))
- Selenium: adding new cases that covered Admin UI page [#4328](https://github.com/jupyterhub/jupyterhub/pull/4328) ([@mouse1203](https://github.com/mouse1203))
- also ignore sqlite backups [#4327](https://github.com/jupyterhub/jupyterhub/pull/4327) ([@minrk](https://github.com/minrk))
- pre-commit: bump isort [#4325](https://github.com/jupyterhub/jupyterhub/pull/4325) ([@minrk](https://github.com/minrk))
- Remove no longer relevant notice in readme [#4324](https://github.com/jupyterhub/jupyterhub/pull/4324) ([@consideRatio](https://github.com/consideRatio))
- Fix the oauthenticator docs api links [#4304](https://github.com/jupyterhub/jupyterhub/pull/4304) ([@GeorgianaElena](https://github.com/GeorgianaElena))
- Selenium testing: adding new case covered the authorisation page [#4298](https://github.com/jupyterhub/jupyterhub/pull/4298) ([@mouse1203](https://github.com/mouse1203))
- docs: fix linkcheck in gallery [#4297](https://github.com/jupyterhub/jupyterhub/pull/4297) ([@minrk](https://github.com/minrk))
- Refactored selenium tests for improved readability [#4278](https://github.com/jupyterhub/jupyterhub/pull/4278) ([@mouse1203](https://github.com/mouse1203))
- remove deprecated import of pipes.quote [#4273](https://github.com/jupyterhub/jupyterhub/pull/4273) ([@minrk](https://github.com/minrk))
- only run testing config on localhost [#4271](https://github.com/jupyterhub/jupyterhub/pull/4271) ([@minrk](https://github.com/minrk))
- pre-commit: autoupdate monthly [#4268](https://github.com/jupyterhub/jupyterhub/pull/4268) ([@minrk](https://github.com/minrk))
- remove unnecessary actions for firefox/geckodriver [#4264](https://github.com/jupyterhub/jupyterhub/pull/4264) ([@minrk](https://github.com/minrk))
- docs: refresh Makefile/make.bat [#4256](https://github.com/jupyterhub/jupyterhub/pull/4256) ([@consideRatio](https://github.com/consideRatio))
- Test docs, links on CI [#4251](https://github.com/jupyterhub/jupyterhub/pull/4251) ([@minrk](https://github.com/minrk))
- maint: fix detail when removing support for py36 [#4248](https://github.com/jupyterhub/jupyterhub/pull/4248) ([@consideRatio](https://github.com/consideRatio))

## Documentation improvements

- fix a couple ref links in changelog [#4334](https://github.com/jupyterhub/jupyterhub/pull/4334) ([@minrk](https://github.com/minrk))
- Backport PR #4316 on branch 3.x (changelog for 3.1.1) [#4318](https://github.com/jupyterhub/jupyterhub/pull/4318) ([@meeseeksmachine](https://github.com/meeseeksmachine))
- changelog for 3.1.1 [#4316](https://github.com/jupyterhub/jupyterhub/pull/4316) ([@minrk](https://github.com/minrk))
- number agreement in authenticators-users-basics [#4309](https://github.com/jupyterhub/jupyterhub/pull/4309) ([@TaofeeqatDev](https://github.com/TaofeeqatDev))
- transferred docs to the FAQ folder [#4307](https://github.com/jupyterhub/jupyterhub/pull/4307) ([@alwasega](https://github.com/alwasega))
- expand database docs [#4292](https://github.com/jupyterhub/jupyterhub/pull/4292) ([@minrk](https://github.com/minrk))
- Document JUPYTER_PREFER_ENV_PATH=0 for shared user environments [#4269](https://github.com/jupyterhub/jupyterhub/pull/4269) ([@minrk](https://github.com/minrk))
- set max depth on api/index toctree [#4259](https://github.com/jupyterhub/jupyterhub/pull/4259) ([@minrk](https://github.com/minrk))
- fix bracket typo in capacity figures [#4250](https://github.com/jupyterhub/jupyterhub/pull/4250) ([@minrk](https://github.com/minrk))
- convert remaining rst files to myst [#4249](https://github.com/jupyterhub/jupyterhub/pull/4249) ([@minrk](https://github.com/minrk))
- doc: fix formatting of spawner env-vars [#4245](https://github.com/jupyterhub/jupyterhub/pull/4245) ([@manics](https://github.com/manics))

## API and Breaking Changes

- Use XSRF tokens for cross-site checks [#4032](https://github.com/jupyterhub/jupyterhub/pull/4032) ([@minrk](https://github.com/minrk))

## Other merged PRs

- Restructured references section of the docs [#4343](https://github.com/jupyterhub/jupyterhub/pull/4343) ([@alwasega](https://github.com/alwasega))
- [pre-commit.ci] pre-commit autoupdate [#4342](https://github.com/jupyterhub/jupyterhub/pull/4342) ([@pre-commit-ci](https://github.com/pre-commit-ci))
- Moved Explanation/Background files [#4340](https://github.com/jupyterhub/jupyterhub/pull/4340) ([@alwasega](https://github.com/alwasega))
- build(deps): bump docker/build-push-action from 3.3.0 to 4.0.0 [#4339](https://github.com/jupyterhub/jupyterhub/pull/4339) ([@dependabot](https://github.com/dependabot))
- Moved last set of Tutorials [#4338](https://github.com/jupyterhub/jupyterhub/pull/4338) ([@alwasega](https://github.com/alwasega))
- simplify some async fixtures [#4332](https://github.com/jupyterhub/jupyterhub/pull/4332) ([@minrk](https://github.com/minrk))
- Added rediraffe using auto redirect builder [#4331](https://github.com/jupyterhub/jupyterhub/pull/4331) ([@alwasega](https://github.com/alwasega))
- Improve contrast on muted text [#4326](https://github.com/jupyterhub/jupyterhub/pull/4326) ([@bl-aire](https://github.com/bl-aire))
- build(deps): bump docker/setup-buildx-action from 2.2.1 to 2.3.0 (now v2) [#4322](https://github.com/jupyterhub/jupyterhub/pull/4322) ([@dependabot](https://github.com/dependabot))
- require sqlalchemy 1.4 [#4319](https://github.com/jupyterhub/jupyterhub/pull/4319) ([@minrk](https://github.com/minrk))
- Moved second half of HowTo documentation [#4314](https://github.com/jupyterhub/jupyterhub/pull/4314) ([@alwasega](https://github.com/alwasega))
- Moved first half of HowTo documentation [#4311](https://github.com/jupyterhub/jupyterhub/pull/4311) ([@alwasega](https://github.com/alwasega))
- Added docs to the  folder [#4305](https://github.com/jupyterhub/jupyterhub/pull/4305) ([@alwasega](https://github.com/alwasega))
- Created folders to house the restructured documentation [#4301](https://github.com/jupyterhub/jupyterhub/pull/4301) ([@alwasega](https://github.com/alwasega))
- build(deps): bump docker/build-push-action from 3.2.0 to 3.3.0 [#4300](https://github.com/jupyterhub/jupyterhub/pull/4300) ([@dependabot](https://github.com/dependabot))
- build(deps): bump json5 from 2.2.1 to 2.2.3 in /jsx [#4291](https://github.com/jupyterhub/jupyterhub/pull/4291) ([@dependabot](https://github.com/dependabot))
- Fix skipped heading level across pages [#4290](https://github.com/jupyterhub/jupyterhub/pull/4290) ([@bl-aire](https://github.com/bl-aire))
- added note on `Spawner.name_template` setting [#4288](https://github.com/jupyterhub/jupyterhub/pull/4288) ([@stevejpurves](https://github.com/stevejpurves))
- [pre-commit.ci] pre-commit autoupdate [#4286](https://github.com/jupyterhub/jupyterhub/pull/4286) ([@pre-commit-ci](https://github.com/pre-commit-ci))
- Fix reoccurring accessibility issues in JupyterHub's pages [#4274](https://github.com/jupyterhub/jupyterhub/pull/4274) ([@bl-aire](https://github.com/bl-aire))
- [pre-commit.ci] pre-commit autoupdate [#4267](https://github.com/jupyterhub/jupyterhub/pull/4267) ([@pre-commit-ci](https://github.com/pre-commit-ci))
- build(deps): bump dessant/support-requests from 2 to 3 [#4263](https://github.com/jupyterhub/jupyterhub/pull/4263) ([@dependabot](https://github.com/dependabot))
- [pre-commit.ci] pre-commit autoupdate [#4247](https://github.com/jupyterhub/jupyterhub/pull/4247) ([@pre-commit-ci](https://github.com/pre-commit-ci))
- build(deps): bump decode-uri-component from 0.2.0 to 0.2.2 in /jsx [#4246](https://github.com/jupyterhub/jupyterhub/pull/4246) ([@dependabot](https://github.com/dependabot))
- more selenium test cases [#4207](https://github.com/jupyterhub/jupyterhub/pull/4207) ([@mouse1203](https://github.com/mouse1203))

## Contributors to this release

([GitHub contributors page for this release](https://github.com/jupyterhub/jupyterhub/graphs/contributors?from=2022-12-05&to=2023-02-10&type=c))

[@ajcollett](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Aajcollett+updated%3A2022-12-05..2023-02-10&type=Issues) | [@alwasega](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Aalwasega+updated%3A2022-12-05..2023-02-10&type=Issues) | [@bl-aire](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Abl-aire+updated%3A2022-12-05..2023-02-10&type=Issues) | [@consideRatio](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3AconsideRatio+updated%3A2022-12-05..2023-02-10&type=Issues) | [@dependabot](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Adependabot+updated%3A2022-12-05..2023-02-10&type=Issues) | [@GeorgianaElena](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3AGeorgianaElena+updated%3A2022-12-05..2023-02-10&type=Issues) | [@github-actions](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Agithub-actions+updated%3A2022-12-05..2023-02-10&type=Issues) | [@julietKiloRomeo](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3AjulietKiloRomeo+updated%3A2022-12-05..2023-02-10&type=Issues) | [@manics](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Amanics+updated%3A2022-12-05..2023-02-10&type=Issues) | [@meeseeksmachine](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Ameeseeksmachine+updated%3A2022-12-05..2023-02-10&type=Issues) | [@minrk](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Aminrk+updated%3A2022-12-05..2023-02-10&type=Issues) | [@mouse1203](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Amouse1203+updated%3A2022-12-05..2023-02-10&type=Issues) | [@pre-commit-ci](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Apre-commit-ci+updated%3A2022-12-05..2023-02-10&type=Issues) | [@sgibson91](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Asgibson91+updated%3A2022-12-05..2023-02-10&type=Issues) | [@Sheila-nk](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3ASheila-nk+updated%3A2022-12-05..2023-02-10&type=Issues) | [@stevejpurves](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Astevejpurves+updated%3A2022-12-05..2023-02-10&type=Issues) | [@TaofeeqatDev](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3ATaofeeqatDev+updated%3A2022-12-05..2023-02-10&type=Issues) | [@vladfreeze](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Avladfreeze+updated%3A2022-12-05..2023-02-10&type=Issues) | [@welcome](https://github.com/search?q=repo%3Ajupyterhub%2Fjupyterhub+involves%3Awelcome+updated%3A2022-12-05..2023-02-10&type=Issues)
```

**Comment by minrk** [comment]
It'll be a small release for a major version (most listed changes have been backported to 3.1, so it's really just #3651 and #4032 so far), but we can make it since that CORS issue keeps coming up and is annoying. I wanted to get runtime-created roles in this release as my next major effort, but I'll settle for blocking only on the single-user extension in #3888 (and making it default, which I'll do in one more PR).

I think we should also wait to do a release for the docs reorg to finish (@alwasega and @sgibson91), since a doc release with partially reorganized docs might be confusing. But we can definitely shift gears into 'preparing a release' mode. I think once #3888 and the default switch PRs land, we can make a prerelease.

**Comment by sgibson91** [comment]
The restructuring of the documentation is very nearly complete. There are some follow-up todos that Allan won't have time to complete during his internship tracked in https://github.com/jupyterhub/jupyterhub/issues/4364

**Comment by minrk** [comment]
https://github.com/jupyterhub/jupyterhub/pull/4375 adds a changelog for 4.0, after which I can publish the first beta.
